File: 386-lexicographical-numbers/lexicographical-numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        res = []

        def generate(curr, limit):
            if curr > limit:
                return
            
            res.append(curr)

            for next_digit in range(10):
                next_num = curr * 10 + next_digit

                if next_num <= limit:
                    generate(next_num, limit)
                else:
                    break
        
        for start in range(1, 10):
            generate(start, n)
        
        return res

File: 386-lexicographical-numbers/lexicographical-numbers.cpp
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> res(n); 
        int curr = 1;

        for (unsigned int i = 0; i < n; ++i) {
            res[i] = curr;

            if (curr * 10 <= n) {
                curr *= 10;
            } else {
                while (curr % 10 == 9 || curr >= n) {
                    curr = int(curr / 10);
                }
                curr += 1;
            }
        }

        return res;
    }
};